#w2step2 - 부하 테스트

Jeongmin94 opened the pull request:
2단계 부하 테스트 미션입니다.

정적 컨텐츠 파일의 압축, 자바스크립트의 비동기 로딩, 캐시 컨트롤 헤더 등의 사용으로 웹 성능을 개선하고, was 소스코드에 존재하는 성능 저하를 유발하는 코드를 제거하고 부하 테스트를 진행했습니다.

소스 코드에 이상한 부분이 있을거라는 생각을 전혀 안하고 진행했다가, 부하 테스트에서 사용할 기능이 정상적으로 작동하지 않아 뭐가 잘못 됐는지 찾는데 시간을 많이 소모한 것 같네요. 알고 보니 강의 자료에 힌트를 주셨는데, 강의 자료를 더 자세히 봐야할 것 같습니다 ㅎㅎ

Jeongmin94 committed:
#w2step1 - 설정 파일 수정

Jeongmin94 committed:
#w1step2 - 웹 성능 예산 측정

Jeongmin94 committed:
#w1step2 - sleep 쿼리 제거(테스트 실패 원인)

Jeongmin94 committed:
#w2step1 - 웹 성능 예산 및 서머 목표 응답시간 추가

Jeongmin94 committed:
Merge branch 'step1'

Jeongmin94 committed:
#w2step2 - 함정 코드 제거

Jeongmin94 committed:
#w2step2 - 함정 코드 제거

Jeongmin94 committed:
#w2step2 - 서빙하지 않는 정적 파일 제거

Jeongmin94 committed:
Revert "#w2step2 - 서빙하지 않는 정적 파일 제거"

This reverts commit ce9c9aa960634fd91c6850731b230c78f643b378.

Jeongmin94 committed:
#w2step2 - spring boot response body compression 설정 추가

Jeongmin94 committed:
#w2step2 - gzip 압축 최소 크기 수정

Jeongmin94 committed:
Merge branch 'main' into step2

Jeongmin94 committed:
#w2step2 - 임베디드 톰캣 캐시 ttl 설정 추가

Jeongmin94 committed:
#w2step2 - 인덱스 페이지의 js 파일을 비동기 처리로 로딩하도록 변경

Jeongmin94 committed:
#w2step2 - MvcConfig 추가

- HTTP 캐시 정책 사용

Jeongmin94 committed:
#w2step2 - MvcConfig 리소스 핸들러 경로 수정

Jeongmin94 committed:
#w2step2 - mvc config 설정 변경

Jeongmin94 committed:
#w2step2 - mvc config 수정

Jeongmin94 committed:
Merge branch 'main' into step2

Jeongmin94 committed:
#w2step2 - 부하테스트 전제조건 추가

Jeongmin94 committed:
#w2step2 - smoke, load, stress 테스트 스크립트 추가 및 테스트 결과 추가

brainbackdoor commented on README.md:
````diff
@@ -41,15 +41,92 @@ npm run dev
+- 목표 TTI 기준인 2.2초 내로 모든 응답을 처리할 수 있어야 함
+
+```bash
+# 웹 성능 개선된 목록
+- 정적 파일 gzip 압축 적용
+- HTTP cache 적용
+- js 파일 비동기 로딩 적용
+
+# WAS
+- 동기 처리 코드 제거
+- 무한 병렬 처리 코드 제거
````
이번 단계에서는 개선까지 할 필요는 없어요.
먼저 테스트를 해서 현재 상태를 계측해보고, 개선 후 다시 테스트를 해서 비교해보는게 중요합니다~

brainbackdoor commented on README.md:
````diff
@@ -41,15 +41,92 @@ npm run dev
+
+```
+R = 3
+http_req_duration = 0.5
````
목표 응답시간 500ms 는 어디서 도출된 지표일까요?
위에서 서버 모든 응답시간의 목표를 2.2s 로 잡지 않았나요?

Jeongmin94 replied:
http_req_duration 값을 어떻게 설정할지 고민하다가 잘 모르겠어서 강의 자료에 있는 500ms로 설정했는데, TTI로 미리 정했다는 것을 잊고 있었네요. TTI 값으로 수정했습니다.

brainbackdoor commented on load/paths/load.js:
```diff
@@ -0,0 +1,40 @@
+        "get stations check": (resp) => resp.status === 200
+    });
+
+    let source = Math.floor(Math.random() * 10 + 1);
+    let target = Math.floor(Math.random() * 10 + 1);
+    let findPath = http.get(`${BASE_URL}/paths/?source=${source}&target=${target}`)
```
👍🏻

brainbackdoor commented on load/paths/load.js:
```diff
@@ -0,0 +1,40 @@
+        "index page check": (resp) => resp.status === 200
+    });
+
+    let stations = http.get(`${BASE_URL}/stations`, params)
```
경로 검색 페이지로 인입하는 url은 `${BASE_URL}/path` 이지 않나요

Jeongmin94 replied:
`/path`로 들어가면 `/stations` 요청이 같이 나가서 착각을 했나보네요. 수정했습니다.

brainbackdoor commented on load/paths/load.js:
```diff
@@ -0,0 +1,40 @@
+import {check} from 'k6';
+
+export let options = {
+    stages: [
+        {duration: '1m', target: 30}
+    ],
```
```suggestion
    stages: [
        {duration: '20s', target: 30},    
        {duration: '1m', target: 30}
        {duration: '20s', target: 0}        
    ],
```

이렇게 구성하면 1분간 VUser를 30까지 올리지 않나요
위와 같이 작성해야 하지 않을까요

brainbackdoor commented on load/paths/smoke.js:
```diff
@@ -0,0 +1,40 @@
+import {check} from 'k6';
+
+export let options = {
+    stages: [
+        {duration: '1m', target: 30}
+    ],
```
smoke 테스트인데 VUser가 30이네요..?

brainbackdoor commented on load/paths/stress.js:
```diff
@@ -0,0 +1,43 @@
+import {check} from 'k6';
+
+export let options = {
+    stages: [
+        {duration: '1m', target: 30},
+        {duration: '1m', target: 55},
+        {duration: '1m', target: 80},
+        {duration: '1m', target: 100}
+    ],
```
부하테스트 스크립트에 남겨둔 피드백을 바탕으로 이 부분도 개선해보세요~

brainbackdoor commented on load/result/result.md:
```diff
@@ -0,0 +1,100 @@
+
+
+running (1m00.2s), 0/5 VUs, 515 complete and 0 interrupted iterations
+default ✓ [======================================] 0/5 VUs  1m0s
```
VUser 값을 평소 VUser값인 6에 근사하게 테스트하셨네요?

brainbackdoor reviewed (changes requested):
2단계 미션 잘 진행하셨네요! 👍🏻 
몇가지 확인할 부분이 있어 Comment 남겼어요~ 확인 바래요~ 🙏🏻

Jeongmin94 committed:
#w2step2 - 리뷰 반영

- http_req_duration 값을 목표 TTI 값으로 변경 및 vuser 값 변경
- 부하 테스트 시나리오에서 vuser 값 변경
- 부하 테스트 시나리오에서 '경로 검색' 페이지로 이동하는 경로에 대한 URL 수정
- 부하 테스트 시나리오에서 smoke, load, stress 테스트에 맞게 stage 값 설정

Jeongmin94 commented:
> 

이번주차에 부하 테스트와 함께 성능 개선을 해야되는 줄 알았는데, 지금 보니 3,4주차에 성능 개선에 대한 작업을 진행하게 되네요!

리뷰해주신 부분들 수정해서 반영했습니다. 감사합니다.

brainbackdoor reviewed (approved):
피드백 반영 확인했어요~ 2단계 미션 진행하느라 고생하셨어요~
주일 잘 보내세요~ 🙏🏻
